server/engine/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java

```diff
@@ -1241,7 +1241,12 @@ public List<Node> visitProcedureDivisionGivingClause(ProcedureDivisionGivingClau
     DataNameContext dt = ctx.dataName();
     Optional<Locality> locality = retrieveLocality(ctx, extendedDocument, copybooks);
     if (dt == null || !locality.isPresent()) return ImmutableList.of();
-    return ImmutableList.of(new ProcedureDivisionReturningNode(locality.get(), extractNameAndLocality(dt)));
+    VariableNameAndLocality nameAndLocality = extractNameAndLocality(dt);
+    ProcedureDivisionReturningNode returningNode = new ProcedureDivisionReturningNode(locality.get(), nameAndLocality);
+    VariableUsageNode variableUsageNode = new VariableUsageNode(nameAndLocality.getName(), nameAndLocality.getLocality());
+    QualifiedReferenceNode qualifiedReferenceNode = new QualifiedReferenceNode(nameAndLocality.getLocality());
+    qualifiedReferenceNode.addChild(variableUsageNode);
+    return ImmutableList.of(returningNode, qualifiedReferenceNode);
```

Comment:
Would it make sense to have the `qualifiedReferenceNode` as a child node of the `returningNode`?